lib/options_validator.js

```diff
@@ -49,29 +49,45 @@ function validationFunction(optionsSchema, providedOptions, validationOptions) {
     return Object.freeze(verifiedOptions);
   }
 
-  Object.keys(verifiedOptions).forEach(optionName => {
-    if (optionsSchema[optionName] && optionsSchema[optionName].type != null) {
-      const optionType = optionsSchema[optionName].type;
+  Object.keys(optionsSchema).forEach(optionName => {
+    const optionFromSchema = optionsSchema[optionName];
+
+    if (optionFromSchema.required && !verifiedOptions.hasOwnProperty(optionName)) {
+      const invalidateMessageRequired = `required option [${optionName}] was not found.`;
+      invalidate(validationLevel, invalidateMessageRequired, logger);
+    }
+
+    if (optionFromSchema.default && !verifiedOptions.hasOwnProperty(optionName)) {
+      verifiedOptions[optionName] = optionFromSchema.default;
+    }
+
+    if (verifiedOptions[optionName] && optionFromSchema.type != null) {
+      const optionType = optionFromSchema.type;
       const optionValue = verifiedOptions[optionName];
-      const invalidateMessage = `${optionName} should be of type ${optionType}, but is of type ${typeof optionValue}.`;
+      const invalidateMessageType = `${optionName} should be of type ${optionType}, but is of type ${typeof optionValue}.`;
 
       if (typeof optionType === 'string') {
         if (
           (optionType === 'object' && !isObject(optionValue)) ||
           typeof optionValue !== optionType
         ) {
-          invalidate(validationLevel, invalidateMessage, logger);
+          invalidate(validationLevel, invalidateMessageType, logger);
         }
       } else if (Array.isArray(optionType)) {
         if (optionType.indexOf(typeof optionValue) === -1) {
-          invalidate(validationLevel, invalidateMessage, logger);
+          invalidate(validationLevel, invalidateMessageType, logger);
         }
       } else {
         if (!(optionValue instanceof optionType)) {
-          invalidate(validationLevel, invalidateMessage, logger);
+          invalidate(validationLevel, invalidateMessageType, logger);
         }
       }
     }
+
+    if (optionFromSchema.deprecated && verifiedOptions.hasOwnProperty(optionName)) {
+      const deprecationMessage = `option [${optionName}] is deprecated and will be removed in a later version.`;
+      invalidate(VALIDATION_LEVEL_WARN, deprecationMessage, logger);
```

Comment:
I think that deprecation should always show the deprecation notice, but should not throw an error, regardless of the provided `validationLevel`. Do you agree, @daprahamian and @mbroadst?

Reply:
I think deprecation notices should conform to the way that `util.deprecate` works. Specifically, we should use [this function](https://github.com/mongodb/node-mongodb-native/blob/master/lib/utils.js#L620)